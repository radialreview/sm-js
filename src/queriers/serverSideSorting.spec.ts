import {
  DEFAULT_TOKEN_NAME,
  DocumentNode,
  EPaginationFilteringSortingInstance,
  MMGQL,
  queryDefinition,
} from '..';
import {
  getMockConfig,
  createMockQueryDefinitions,
  getPrettyPrintedGQL,
  generateUserNode,
  createMockDataItems,
  mockUserData,
} from '../specUtilities';

test(`query gql matches the expectation when specifying a sorting param`, async () => {
  const { mmGQLInstance } = setupTest({
    mockData: {
      users: createMockDataItems({
        sampleMockData: mockUserData,
        items: [],
      }),
    },
    onQueryPerformed: query => {
      expect(getPrettyPrintedGQL(query)).toMatchSnapshot();
    },
  });

  await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ firstName }) => ({
        firstName,
      }),
      sort: {
        firstName: 'asc',
      },
    }),
  });
});

test(`query gql matches the expectation when specifying a sort param for a relational property`, async () => {
  const { mmGQLInstance } = setupTest({
    mockData: {
      users: createMockDataItems({
        sampleMockData: mockUserData,
        items: [],
      }),
    },
    onQueryPerformed: query => {
      expect(getPrettyPrintedGQL(query)).toMatchSnapshot();
    },
  });

  await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ firstName, todos }) => ({
        firstName,
        todos: todos({
          map: ({ task }) => ({ task }),
        }),
      }),
      sort: {
        todos: {
          task: 'asc',
        },
      },
    }),
  });
});

test(`query gql matches the expectation when specifying a sort params with priority`, async () => {
  const { mmGQLInstance } = setupTest({
    mockData: {
      users: createMockDataItems({
        sampleMockData: mockUserData,
        items: [],
      }),
    },
    onQueryPerformed: query => {
      expect(getPrettyPrintedGQL(query)).toMatchSnapshot();
    },
  });

  await mmGQLInstance.query({
    users: queryDefinition({
      def: generateUserNode(mmGQLInstance),
      map: ({ firstName, lastName }) => ({
        firstName,
        lastName,
      }),
      sort: {
        firstName: {
          priority: 1,
          direction: 'asc',
        },
        lastName: {
          priority: 2,
          direction: 'desc',
        },
      },
    }),
  });
});

function setupTest(opts: {
  mockData?: any;
  onQueryPerformed: (query: DocumentNode) => void;
}) {
  const mmGQLInstance = new MMGQL(
    getMockConfig({
      mockData: opts.mockData,
      onQueryPerformed: opts.onQueryPerformed,
      generateMockData: false,
      paginationFilteringSortingInstance:
        EPaginationFilteringSortingInstance.SERVER,
    })
  );
  mmGQLInstance.setToken({
    tokenName: DEFAULT_TOKEN_NAME,
    token: 'mock token',
  });
  const queryDefinitions = createMockQueryDefinitions(mmGQLInstance);

  return { mmGQLInstance, queryDefinitions, createMockQueryDefinitions };
}
