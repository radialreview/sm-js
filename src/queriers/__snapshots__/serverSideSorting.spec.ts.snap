// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query gql matches the expectation when specifying a sort param for a relational property 1`] = `
"query query1 {
 users: users(order: [{todos: DESC}]) {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
     todos: todos {
       nodes {
         id
         version
         lastUpdatedBy
         type
         task
       }
       pageInfo {
         endCursor
         startCursor
         hasNextPage
         hasPreviousPage
       }
     }
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query gql matches the expectation when specifying a sort params with priority 1`] = `
"query query2 {
 users: users(order: [{firstName: ASC}, {lastName: DESC}]) {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
     lastName
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;

exports[`query gql matches the expectation when specifying a sorting param 1`] = `
"query query0 {
 users: users(order: [{firstName: ASC}]) {
   nodes {
     id
     version
     lastUpdatedBy
     type
     firstName
   }
   pageInfo {
     endCursor
     startCursor
     hasNextPage
     hasPreviousPage
   }
 }
}"
`;
